numba_dpex/kernel_api/local_accessor.py

```diff
@@ -0,0 +1,101 @@
+# SPDX-FileCopyrightText: 2023 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Implements a Python analogue to SYCL's local_accessor class. The class is
+intended to be used in pure Python code when prototyping a kernel function
+and to be passed to an actual kernel function for local memory allocation.
+"""
+import numpy
+
+
+class LocalAccessor:
+    """
+    The ``LocalAccessor`` class is analogous to SYCL's ``local_accessor``
+    class. The class acts a s proxy to allocating device local memory and
+    accessing that memory from within a :func:`numba_dpex.kernel` decorated
+    function.
+    """
+
+    def _verify_positive_integral_list(self, ls):
+        """Checks if all members of a list are positive integers."""
+
+        ret = False
+        try:
+            ret = all(int(val) > 0 for val in ls)
+        except ValueError:
+            pass
+
+        return ret
+
+    def __init__(self, shape, dtype) -> None:
+        """Creates a new LocalAccessor instance of the given shape and dtype."""
+
+        if not isinstance(shape, (list, tuple)):
+            if hasattr(shape, "tolist"):
+                fn = getattr(shape, "tolist")
+                if callable(fn):
+                    self._shape = shape.tolist()
+            else:
+                try:
+                    self._shape = [
+                        shape,
+                    ]
+                except Exception as e:
+                    raise TypeError(
+                        "Argument shape must a non-negative integer, "
+                        "or a list/tuple of such integers."
+                    ) from e
+        else:
+            self._shape = list(shape)
+
+        # Make sure shape is made up a supported types
+        if not self._verify_positive_integral_list(self._shape):
+            raise TypeError(
+                "Argument shape must a non-negative integer, "
+                "or a list/tuple of such integers."
+            )
+
+        # Make sure shape has a rank between (1..3)
+        if len(self._shape) < 1 or len(self._shape) > 3:
+            raise TypeError("LocalAccessor can only have up to 3 dimensions.")
+
+        self._dtype = dtype
+
+        if self._dtype not in [
+            numpy.float32,
+            numpy.float64,
+            numpy.int32,
+            numpy.int64,
+        ]:
+            raise TypeError(
+                f"Argument dtype {dtype} is not supported. numpy.float32, "
+                "numpy.float64, numpy.int32, numpy.int64 are the currently "
+                "supported dtypes."
+            )
+
+        self._data = numpy.empty(self._shape, dtype=self._dtype)
+
+    def __getitem__(self, idx_obj):
+        """Returns the value stored at the position represented by idx_obj in
+        the self._data ndarray.
+        """
+
+        return self._data[idx_obj]
+
+    def __setitem__(self, idx_obj, val):
```

Comment:
Why do you need this?
Any attempt to read/write to local memory on host should result in exception.

Or I misunderstands something?

Reply:
Sorry, I missed your comment.

This class is basically a simulator in pure Python for prototyping and testing. Borrowing idea's for numba-mlir's simulator.

The usage is, a `LocalAccessor` is crated in pure Python and passed as a kernel argument. In simulator mode it is just a NumPy array so works as expected. In compiled mode it will be an actual local array. What do you think?

Reply:
Since there is no examples I'm not sure how it should work, but I assume, that user need to create instance of this class in order to use local memory inside of kernel. Something like that:

```python
loc_mem = LocalAccessor(...)
kernel(..., loc_mem, ...)
```
If that's assumption is correct, then user may try to access local memory outside of kernel:
```python
loc_mem = LocalAccessor(...)
loc_mem[0] = 10
kernel(..., loc_mem, ...)
```

And in this case an exception must be thrown.

Reply:
Let's then introduce `LocalAccessorMock` with those methods implemented.

Reply:
I have made these changes. Can you please review again?